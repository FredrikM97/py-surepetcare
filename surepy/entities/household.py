from surepy.const import API_ENDPOINT_V1
from surepy.const import API_ENDPOINT_V2
from surepy.devices import load_device_class
from surepy.devices.base import SurepyDevice
from surepy.entities.pet import Pet
from surepy.enums import ProductId
from surepy.helper import AbstractHasGet


class HouseholdMixin(AbstractHasGet):
    async def get_household(self, household_id: int):
        return await self.get(f"{API_ENDPOINT_V1}/household/{household_id}")

    async def get_households(self):
        return (await self.get(f"{API_ENDPOINT_V1}/household"))["data"]

    async def get_devices(self, household_ids: list[int]) -> list[SurepyDevice]:
        """Get devices for a list of household IDs."""
        products = set(ProductId)

        devices = []
        for household_id in household_ids:
            household_devices = (
                await self.get(f"{API_ENDPOINT_V1}/device", params={"HouseholdId": household_id})
            )["data"]
            for device in household_devices:
                if device["product_id"] in products:
                    devices.append(load_device_class(ProductId(device["product_id"]))(self, device))

        return devices

    async def get_product(self, product_id: ProductId, device_id: int):
        return await self.get(f"{API_ENDPOINT_V2}/product/{product_id}/device/{device_id}/control")

    async def get_pets(self, household_id: int) -> list[Pet]:
        pets = []
        for pet in (await self.get(f"{API_ENDPOINT_V1}/pet", params={"HouseholdId": household_id}))["data"]:
            pets.append(Pet(self, pet))
        return pets
