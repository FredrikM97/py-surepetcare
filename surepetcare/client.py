import logging

from surepetcare.command import Command
from surepetcare.entities.api import ApiResponse
from surepetcare.security.auth import AuthClient

logger = logging.getLogger(__name__)


class SurePetcareClient(AuthClient, ApiResponse):
    async def get(self, endpoint: str, params: dict | None = None, headers=None) -> dict:
        if not headers:
            headers = self._generate_headers(self.token)
        await self.set_session()
        async with self.session.get(
            endpoint, params=params, headers=headers
        ) as response:
            if hasattr(self, 'populate_headers'):
                self.populate_headers(response)
            if not response.ok:
                print(response)
                raise Exception(f"Error {endpoint} {response.status}: {await response.text()}")
            if response.status == 204:
                logger.info(f"GET {endpoint} returned 204 No Content")
                return {}
            return await response.json()

    async def post(self, endpoint: str, data: dict | None = None, headers=None) -> dict:
        if not headers:
            headers = self._generate_headers()
        await self.set_session()
        async with self.session.post(
            endpoint, json=data, headers=headers
        ) as response:
            if hasattr(self, 'populate_headers'):
                self.populate_headers(response)
            if not response.ok:
                raise Exception(f"Error {response.status}: {await response.text()}")
            if response.status == 204:
                logger.info(f"POST {endpoint} returned 204 No Content")
                return {}
            return await response.json()

    async def api(self, command: Command):
        method = command.method.lower()
        if method == "get":
            coro = self.get(
                command.endpoint, 
                params=command.params
                )
        elif method == "post":
            coro = self.post(
                command.endpoint, 
                data=command.params
            )

        else:
            raise NotImplementedError(f"HTTP method {command.method} not supported.")
        response = await coro
        if command.callback:
            return command.callback(response)
        return response