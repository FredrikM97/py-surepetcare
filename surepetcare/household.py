import logging
from surepetcare.command import Command
from surepetcare.const import API_ENDPOINT_PRODUCTION, API_ENDPOINT_V1
from surepetcare.devices import load_device_class
from surepetcare.devices.pet import Pet
from surepetcare.enums import ProductId

logger = logging.getLogger(__name__)

class Household:
    def __init__(self, data: dict):
        self.data = data
        self.id = data["id"]
        # Add other fields as needed

    def get_pets(self):
        def parse(response):
            if response['status'] == 304:
                return self.data.get("pets", [])
            pets = [Pet(p) for p in response["data"]]
            self.data['pets'] = pets
            return pets
            

        return Command(
            method="GET", endpoint=f"{API_ENDPOINT_V1}/pet", params={"HouseholdId": self.id}, callback=parse
        )

    def get_devices(self):
        def parse(response):
            logger.info(f"Parsing devices for household {self.id}: {response}")
            if response['status'] == 304:
                logger.info("Returning cached devices")
                return self.data.get("devices",[])
            devices = []
            for device in response["data"]:
                if device["product_id"] in set(ProductId):
                    devices.append(load_device_class(device["product_id"])(device))
            self.data['devices'] = devices
            return devices

        return Command(
            method="GET",
            endpoint=f"{API_ENDPOINT_PRODUCTION}/device",
            params={"HouseholdId": self.id},
            callback=parse,
        )

    @staticmethod
    def get_households():
        def parse(response):
            return [Household(h) for h in response["data"]]

        return Command(method="GET", endpoint=f"{API_ENDPOINT_PRODUCTION}/household", params={}, callback=parse, reuse=False)

    @staticmethod
    def get_household(household_id: int):
        return Command(method="GET", endpoint=f"{API_ENDPOINT_PRODUCTION}/household/{household_id}", reuse=False)

    @staticmethod
    def get_product(product_id: ProductId, device_id: int):
        """TODO: Move to devices instead"""
        return Command(
            method="GET", endpoint=f"{API_ENDPOINT_PRODUCTION}/product/{product_id}/device/{device_id}/control", reuse=False
        )
