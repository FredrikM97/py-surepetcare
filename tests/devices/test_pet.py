import json

import pytest

from surepetcare.household import Household
from tests.mock_helpers import MockClient
from tests.mock_helpers import recursive_dump


@pytest.fixture
def device_file():
    return "tests/fixture/pet.json"


@pytest.fixture
def household_file():
    return "tests/fixture/household.json"


@pytest.mark.asyncio
async def test_snapshot(snapshot, household_file, device_file):
    # snapshot.snapshot_dir = "tests/snapshots"
    client = MockClient(fixture_file=device_file)
    client.set_mock_response(household_file)
    household = await client.api(Household.get_household(7777))
    client.reset()
    pet = (await client.api(household.get_pets()))[0]
    await client.api(pet.refresh())
    result = json.dumps(recursive_dump(pet), indent=2)
    assert result == snapshot
