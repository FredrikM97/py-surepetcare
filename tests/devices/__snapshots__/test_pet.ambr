# serializer version: 1
# name: test_snapshot
  dict({
    'control': dict({
    }),
    'entity_info': dict({
      'household_id': 7777,
      'id': 123455,
      'name': 'Maui',
      'photo': dict({
        'created_at': datetime.datetime(2023, 10, 3, 15, 18, 9, tzinfo=TzInfo(UTC)),
        'hash': 'UCcarrDHzX0MkjEORQqCXNLKxsdweCQPJjl3OQ7s',
        'id': 623733,
        'location': 'https://userimages.production.surehub.io/thm/3139194472/UCcarrDHzX0MkjEORQqCXNLKxsdweCQPJjl3OQ7s.jpg',
        'title': '1696346281223-cropped.jpg?1696346286094',
        'updated_at': datetime.datetime(2023, 10, 3, 15, 18, 9, tzinfo=TzInfo(UTC)),
        'uploading_user_id': 3139194472,
        'version': 0,
      }),
      'product_id': 0,
      'tag': dict({
        'created_at': '2019-09-04T20:28:11+00:00',
        'id': 60978,
        'supported_product_ids': list([
          3,
          4,
          6,
          8,
          10,
          32,
        ]),
        'tag': '752.093200030403',
        'updated_at': '2020-03-11T14:21:07+00:00',
        'version': 1,
      }),
      'tag_id': 60978,
    }),
    'last_fetched_datetime': '<ANY>',
    'status': dict({
      'report': dict({
        'drinking': list([
        ]),
        'feeding': list([
          dict({
            'bowl_count': 2,
            'context': 1,
            'device_id': 269654,
            'duration': 44,
            'from_': datetime.datetime(2025, 5, 27, 22, 37, 30, tzinfo=TzInfo(UTC)),
            'to': datetime.datetime(2025, 5, 27, 22, 38, 14, tzinfo=TzInfo(UTC)),
            'weights': list([
              dict({
                'change': -4.0,
                'food_type_id': <FoodType.WET: 1>,
                'index': <BowlPosition.LEFT: 0>,
                'target_weight': 0.0,
                'weight': 24.0,
              }),
              dict({
                'change': 0.0,
                'food_type_id': <FoodType.WET: 1>,
                'index': <BowlPosition.RIGHT: 1>,
                'target_weight': 0.0,
                'weight': 0.0,
              }),
            ]),
          }),
          dict({
            'bowl_count': 1,
            'context': 1,
            'device_id': 1181149,
            'duration': 89,
            'from_': datetime.datetime(2025, 5, 27, 1, 43, 35, tzinfo=TzInfo(UTC)),
            'to': datetime.datetime(2025, 5, 27, 1, 45, 4, tzinfo=TzInfo(UTC)),
            'weights': list([
              dict({
                'change': -4.0,
                'food_type_id': <FoodType.DRY: 2>,
                'index': <BowlPosition.LEFT: 0>,
                'target_weight': 0.0,
                'weight': 71.0,
              }),
              dict({
                'change': 0.0,
                'index': <BowlPosition.RIGHT: 1>,
                'target_weight': 0.0,
                'weight': 0.0,
              }),
            ]),
          }),
          dict({
            'bowl_count': 1,
            'context': 1,
            'device_id': 1181149,
            'duration': 67,
            'from_': datetime.datetime(2025, 5, 27, 3, 50, 44, tzinfo=TzInfo(UTC)),
            'to': datetime.datetime(2025, 5, 27, 3, 51, 51, tzinfo=TzInfo(UTC)),
            'weights': list([
              dict({
                'change': -3.0,
                'food_type_id': <FoodType.DRY: 2>,
                'index': <BowlPosition.LEFT: 0>,
                'target_weight': 0.0,
                'weight': 68.0,
              }),
              dict({
                'change': 0.0,
                'index': <BowlPosition.RIGHT: 1>,
                'target_weight': 0.0,
                'weight': 0.0,
              }),
            ]),
          }),
          dict({
            'bowl_count': 1,
            'context': 1,
            'device_id': 1181149,
            'duration': 85,
            'from_': datetime.datetime(2025, 5, 27, 9, 28, 56, tzinfo=TzInfo(UTC)),
            'to': datetime.datetime(2025, 5, 27, 9, 30, 21, tzinfo=TzInfo(UTC)),
            'weights': list([
              dict({
                'change': -3.0,
                'food_type_id': <FoodType.DRY: 2>,
                'index': <BowlPosition.LEFT: 0>,
                'target_weight': 0.0,
                'weight': 65.0,
              }),
              dict({
                'change': 0.0,
                'index': <BowlPosition.RIGHT: 1>,
                'target_weight': 0.0,
                'weight': 0.0,
              }),
            ]),
          }),
          dict({
            'bowl_count': 1,
            'context': 1,
            'device_id': 1181149,
            'duration': 71,
            'from_': datetime.datetime(2025, 5, 27, 10, 3, 25, tzinfo=TzInfo(UTC)),
            'to': datetime.datetime(2025, 5, 27, 10, 4, 36, tzinfo=TzInfo(UTC)),
            'weights': list([
              dict({
                'change': -2.0,
                'food_type_id': <FoodType.DRY: 2>,
                'index': <BowlPosition.LEFT: 0>,
                'target_weight': 0.0,
                'weight': 63.0,
              }),
              dict({
                'change': 0.0,
                'index': <BowlPosition.RIGHT: 1>,
                'target_weight': 0.0,
                'weight': 0.0,
              }),
            ]),
          }),
          dict({
            'bowl_count': 1,
            'context': 1,
            'device_id': 1181149,
            'duration': 122,
            'from_': datetime.datetime(2025, 5, 27, 16, 21, 43, tzinfo=TzInfo(UTC)),
            'to': datetime.datetime(2025, 5, 27, 16, 23, 45, tzinfo=TzInfo(UTC)),
            'weights': list([
              dict({
                'change': -6.0,
                'food_type_id': <FoodType.DRY: 2>,
                'index': <BowlPosition.LEFT: 0>,
                'target_weight': 0.0,
                'weight': 56.0,
              }),
              dict({
                'change': 0.0,
                'index': <BowlPosition.RIGHT: 1>,
                'target_weight': 0.0,
                'weight': 0.0,
              }),
            ]),
          }),
          dict({
            'bowl_count': 1,
            'context': 1,
            'device_id': 1181149,
            'duration': 80,
            'from_': datetime.datetime(2025, 5, 27, 18, 24, 10, tzinfo=TzInfo(UTC)),
            'to': datetime.datetime(2025, 5, 27, 18, 25, 30, tzinfo=TzInfo(UTC)),
            'weights': list([
              dict({
                'change': -3.0,
                'food_type_id': <FoodType.DRY: 2>,
                'index': <BowlPosition.LEFT: 0>,
                'target_weight': 0.0,
                'weight': 54.0,
              }),
              dict({
                'change': 0.0,
                'index': <BowlPosition.RIGHT: 1>,
                'target_weight': 0.0,
                'weight': 0.0,
              }),
            ]),
          }),
          dict({
            'bowl_count': 1,
            'context': 1,
            'device_id': 1181149,
            'duration': 65,
            'from_': datetime.datetime(2025, 5, 27, 19, 28, 49, tzinfo=TzInfo(UTC)),
            'to': datetime.datetime(2025, 5, 27, 19, 29, 54, tzinfo=TzInfo(UTC)),
            'weights': list([
              dict({
                'change': -2.0,
                'food_type_id': <FoodType.DRY: 2>,
                'index': <BowlPosition.LEFT: 0>,
                'target_weight': 0.0,
                'weight': 51.0,
              }),
              dict({
                'change': 0.0,
                'index': <BowlPosition.RIGHT: 1>,
                'target_weight': 0.0,
                'weight': 0.0,
              }),
            ]),
          }),
          dict({
            'bowl_count': 1,
            'context': 1,
            'device_id': 1181149,
            'duration': 81,
            'from_': datetime.datetime(2025, 5, 27, 20, 24, 56, tzinfo=TzInfo(UTC)),
            'to': datetime.datetime(2025, 5, 27, 20, 26, 17, tzinfo=TzInfo(UTC)),
            'weights': list([
              dict({
                'change': -5.0,
                'food_type_id': <FoodType.DRY: 2>,
                'index': <BowlPosition.LEFT: 0>,
                'target_weight': 0.0,
                'weight': 47.0,
              }),
              dict({
                'change': 0.0,
                'index': <BowlPosition.RIGHT: 1>,
                'target_weight': 0.0,
                'weight': 0.0,
              }),
            ]),
          }),
        ]),
        'movement': list([
        ]),
      }),
    }),
    'timezone': 'Europe/Stockholm',
  })
# ---
# name: test_snapshot.1
  dict({
    'control': dict({
    }),
    'entity_info': dict({
      'household_id': 7777,
      'id': 629768,
      'name': 'Fionna',
      'photo': dict({
        'created_at': datetime.datetime(2025, 2, 28, 15, 34, 12, tzinfo=TzInfo(UTC)),
        'hash': 'R01qBHfrNHNkeMiNQrtfgoWlOtUGj8TqTCzgyUZPsI',
        'id': 820465,
        'location': 'https://userimages.production.surehub.io/thm/XXXXXXXXX/R01qBHfrNHNkeMiNQrtfgoWlOtUGj8TqTCzgyUZPsI.jpg',
        'uploading_user_id': 9999999,
        'version': 0,
      }),
      'product_id': 0,
      'tag': dict({
        'created_at': '2025-02-28T13:34:39+00:00',
        'id': 1725049,
        'supported_product_ids': list([
          3,
          4,
          6,
          8,
          10,
          32,
        ]),
        'tag': '756.098101172569',
        'version': 0,
      }),
      'tag_id': 1725049,
    }),
    'last_fetched_datetime': '<ANY>',
    'status': dict({
      'report': dict({
        'drinking': list([
        ]),
        'feeding': list([
          dict({
            'bowl_count': 1,
            'context': 5,
            'device_id': 1184438,
            'duration': 5324,
            'from_': datetime.datetime(2025, 9, 11, 4, 7, 14, tzinfo=TzInfo(UTC)),
            'to': datetime.datetime(2025, 9, 11, 5, 35, 58, tzinfo=TzInfo(UTC)),
            'weights': list([
              dict({
                'change': 41.0,
                'food_type_id': <FoodType.WET: 1>,
                'index': <BowlPosition.LEFT: 0>,
                'target_weight': 0.0,
                'weight': 120.0,
              }),
              dict({
                'change': 0.0,
                'food_type_id': <FoodType.DRY: 2>,
                'index': <BowlPosition.RIGHT: 1>,
                'target_weight': 0.0,
                'weight': 0.0,
              }),
            ]),
          }),
          dict({
            'bowl_count': 1,
            'context': 1,
            'device_id': 1184438,
            'duration': 221,
            'from_': datetime.datetime(2025, 9, 11, 5, 36, 31, tzinfo=TzInfo(UTC)),
            'to': datetime.datetime(2025, 9, 11, 5, 40, 12, tzinfo=TzInfo(UTC)),
            'weights': list([
              dict({
                'change': -26.0,
                'food_type_id': <FoodType.WET: 1>,
                'index': <BowlPosition.LEFT: 0>,
                'target_weight': 0.0,
                'weight': 92.0,
              }),
              dict({
                'change': 0.0,
                'food_type_id': <FoodType.DRY: 2>,
                'index': <BowlPosition.RIGHT: 1>,
                'target_weight': 0.0,
                'weight': 0.0,
              }),
            ]),
          }),
          dict({
            'bowl_count': 1,
            'context': 1,
            'device_id': 1184438,
            'duration': 149,
            'from_': datetime.datetime(2025, 9, 11, 6, 33, 27, tzinfo=TzInfo(UTC)),
            'to': datetime.datetime(2025, 9, 11, 6, 35, 56, tzinfo=TzInfo(UTC)),
            'weights': list([
              dict({
                'change': -16.0,
                'food_type_id': <FoodType.WET: 1>,
                'index': <BowlPosition.LEFT: 0>,
                'target_weight': 0.0,
                'weight': 76.0,
              }),
              dict({
                'change': 0.0,
                'food_type_id': <FoodType.DRY: 2>,
                'index': <BowlPosition.RIGHT: 1>,
                'target_weight': 0.0,
                'weight': 0.0,
              }),
            ]),
          }),
        ]),
        'movement': list([
          dict({
            'duration': 2954,
            'entry_device_id': 1299999,
            'entry_movement_id': 0,
            'exit_device_id': 1299999,
            'exit_movement_id': 0,
            'from_': datetime.datetime(2025, 9, 11, 5, 44, 2, tzinfo=TzInfo(UTC)),
            'to': datetime.datetime(2025, 9, 11, 6, 33, 16, tzinfo=TzInfo(UTC)),
          }),
          dict({
            'duration': 2869,
            'entry_device_id': 1299999,
            'entry_movement_id': 0,
            'exit_device_id': 1299999,
            'exit_movement_id': 0,
            'from_': datetime.datetime(2025, 9, 11, 6, 43, 51, tzinfo=TzInfo(UTC)),
            'to': datetime.datetime(2025, 9, 11, 7, 31, 40, tzinfo=TzInfo(UTC)),
          }),
          dict({
            'duration': 963,
            'entry_device_id': 1299999,
            'entry_movement_id': 0,
            'exit_device_id': 1299999,
            'exit_movement_id': 0,
            'from_': datetime.datetime(2025, 9, 11, 7, 32, 20, tzinfo=TzInfo(UTC)),
            'to': datetime.datetime(2025, 9, 11, 7, 48, 23, tzinfo=TzInfo(UTC)),
          }),
          dict({
            'active': True,
            'entry_movement_id': 0,
            'exit_movement_id': 0,
            'from_': datetime.datetime(2025, 9, 11, 10, 35, 57, tzinfo=TzInfo(UTC)),
          }),
        ]),
      }),
    }),
    'timezone': 'Europe/Stockholm',
  })
# ---
